pyrival/strings/hashing.py

```diff
@@ -9,21 +9,25 @@ class Hashing:
     def __init__(self, s, mod=HMOD, base1=HBASE1, base2=HBASE2):
         self.mod, self.base1, self.base2 = mod, base1, base2
         self._len = _len = len(s)
-        f_hash, s_hash = [0] * (_len + 1), [0] * (_len + 1)
+        f_hash, f_pw = [0] * (_len + 1), [1] * (_len + 1)
+        s_hash, s_pw = f_hash[:], f_pw[:]
         for i in range(_len):
             f_hash[i + 1] = (base1 * f_hash[i] + s[i]) % mod
             s_hash[i + 1] = (base2 * s_hash[i] + s[i]) % mod
-        self.f_hash, self.s_hash = f_hash, s_hash
+            f_pw[i + 1] = (base1 * f_pw[i]) % mod
+            s_pw[i + 1] = (base2 * s_pw[i]) % mod
+        self.f_hash, self.f_pw = f_hash, f_pw
+        self.s_hash, self.s_pw = s_hash, s_pw
 
     def hashed(self, start, stop):
         return (
-            (self.f_hash[stop] - pow(self.base1, stop - start, self.mod) * self.f_hash[start]) % self.mod,
-            (self.s_hash[stop] - pow(self.base2, stop - start, self.mod) * self.s_hash[start]) % self.mod,
+            (self.f_hash[stop] - self.f_pw[stop - start] * self.f_hash[start]) % self.mod,
+            (self.s_hash[stop] - self.s_pw[stop - start] * self.s_hash[start]) % self.mod,
         )
 
     def get_hashes(self, length):
         mod = self.mod
-        pow_base1, pow_base2 = pow(self.base1, length, mod), pow(self.base2, length, mod)
-        f_hash = [(self.f_hash[i + length] - pow_base1 * self.f_hash[i]) % mod for i in range(self._len - length + 1)]
-        s_hash = [(self.s_hash[i + length] - pow_base2 * self.s_hash[i]) % mod for i in range(self._len - length + 1)]
-        return f_hash, s_hash
+        return (
+            [(self.f_hash[i + length] - self.f_pw[length] * self.f_hash[i]) % mod for i in range(self._len - length + 1)],
+            [(self.s_hash[i + length] - self.s_pw[length] * self.s_hash[i]) % mod for i in range(self._len - length + 1)],
```

Comment:
Maybe replace `mod` with `self.mod`, so that you can remove `mod = self.mod` above.

If my suggested change makes it too slow, then this is fine too.

Reply:
Thanks for the quick review!

My solution doesn't use get_hashes, so I cannot determine the performance implications of inlining the field access.

Reply:
Ah, I see, that's fair.

I think it's alright then, this is usually not an issue on PyPy, but vanilla Python code likes it when the variables are local.

Reply:
That's my general experience too. Should I perform the change under the assumption that people doing CP with Python use PyPy? ;)

Reply:
Sure, I think that is reasonable, thanks!